File: 2023/day03/solution.py
#!/usr/bin/env python3
import sys
import re
from collections import defaultdict

sys.setrecursionlimit(1_000_000)
DEBUG = 0
FILENAME = 'input.txt'
PART1 = PART2 = True
D4 = [(-1, 0), (0, 1), (1, 0), (0, -1)]

for arg in sys.argv[1:]:
  if arg.startswith('-e'): FILENAME = f'''example{arg[2:] if len(arg)>2 else ''}.txt'''
  elif arg.startswith('-d'): DEBUG = int(arg[2:]) if len(arg) > 2 else 1
  else:
    match arg:
      case '-i':
        FILENAME = 'input.txt'
      case 'p1':
        PART1, PART2 = True, False
      case 'p2':
        PART1, PART2 = False, True
      case 'p0':
        PART1, PART2 = False, False
      case _:
        raise Exception(f'unknown {arg=}')

with open(FILENAME, 'r') as f:
  text = f.read().splitlines()


def part1():
  total = 0
  digit_regex = re.compile(r'(\d+)')
  symbol_regex = re.compile(r'([^\d.])')
  rows = len(text)
  cols = len(text[0])
  if DEBUG > 0: print(f'{rows=} {cols=}')
  for row, line in enumerate(text):
    for match in re.finditer(digit_regex, line):
      #m.span is a tuple of start and end char ixs, though end is +1 due to reasons
      col_start = match.span()[0]
      col_end = match.span()[1] - 1
      length = col_end - col_start + 1
      part = match.group(1)
      if DEBUG > 0: print(f'{part=} {row=} {col_start=} {col_end=} {length=}')
      perimeter: list[str] = []
      if row > 0:
        value = text[row - 1][max(0, col_start - 1):min(col_end + 2, cols - 1)]
        if DEBUG > 1: print(f'  above: {value}')
        perimeter += [value]
      if row < rows - 1:
        value = text[row + 1][max(0, col_start - 1):min(col_end + 2, cols - 1)]
        if DEBUG > 1: print(f'  below: {value}')
        perimeter += [value]
      if col_start > 0:
        value = text[row][col_start - 1]
        if DEBUG > 1: print(f'  left: {value}')
        perimeter += [value]
      if col_end < cols - 1:
        value = text[row][col_end + 1]
        if DEBUG > 1: print(f'  right: {value}')
        perimeter += [value]
      found = re.search(symbol_regex, ''.join(perimeter))
      # if DEBUG > 0: print(f'{part=} {row=} {col_start=} {col_end=} {length=} {perimeter=} {found=}')
      if DEBUG > 0: print(f'  {perimeter=} {found=}')
      if found: total += int(part)
      # if row >= 5: exit()
  print(f'part 1: {total=}')


def part2():
  digit_regex = re.compile(r'(\d+)')
  symbol_regex = re.compile(r'(\*)')
  gears: defaultdict[tuple[int, int], list[int]] = defaultdict(list)
  rows = len(text)
  cols = len(text[0])
  if DEBUG > 0: print(f'{rows=} {cols=}')
  for row, line in enumerate(text):
    for match in re.finditer(digit_regex, line):
      #m.span is a tuple of start and end char ixs, though end is +1 due to reasons
      col_start = match.span()[0]
      col_end = match.span()[1] - 1
      length = col_end - col_start + 1
      value = int(match.group(1))
      if DEBUG > 0: print(f'{value=} {row=} {col_start=} {col_end=} {length=}')

      if row > 0:
        string = text[row - 1][max(0, col_start - 1):min(col_end + 2, cols - 1)]
        for gear_match in re.finditer(symbol_regex, string):
          gears[(row - 1, max(0, col_start - 1) + gear_match.span()[0])].append(value)
          if DEBUG > 1: print(f'  gear above at {(row-1,  max(0, col_start - 1)+gear_match.span()[0])}')

      if row < rows - 1:
        string = text[row + 1][max(0, col_start - 1):min(col_end + 2, cols - 1)]
        for gear_match in re.finditer(symbol_regex, string):
          gears[(row + 1, max(0, col_start - 1) + gear_match.span()[0])].append(value)
          if DEBUG > 1: print(f'  gear below at {(row+1,  max(0, col_start - 1)+gear_match.span()[0])}')

      if col_start > 0:
        if text[row][col_start - 1] == '*':
          gears[(row, col_start - 1)].append(value)
          if DEBUG > 1: print(f'  gear left at {(row, col_start - 1)}')

      if col_end < cols - 1:
        if text[row][col_end + 1] == '*':
          gears[(row, col_end + 1)].append(value)
          if DEBUG > 1: print(f'  gear right at {(row, col_end + 1)}')

      # if row >= 5: exit()

  total = sum(v[0] * v[1] for v in gears.values() if len(v) == 2)
  print(f'part 2: {total=}')


if PART1: part1()
if PART2: part2()

File: 2023/day03/solution.ts
#!/usr/bin/env -S deno --allow-read

const DEBUG = Deno.args.reduce((acc, item) => (item.startsWith('-d') ? Number(item.slice(2) || '1') : acc), 0);
const FILENAME = Deno.args.reduce(
  (acc, item) => (item == '-i' ? 'input.txt' : item.startsWith('-e') ? `example${item.slice(2)}.txt` : acc),
  'example.txt'
);
const [PART1, PART2] = Deno.args.reduce(
  (acc, item) => (item == '-p0' ? [false, false] : item == '-p1' ? [true, false] : item == '-p2' ? [false, true] : acc),
  [true, true]
);
console.log({ FILENAME, DEBUG, PART1, PART2 });

// deno-lint-ignore no-explicit-any
const debug = (level: number, ...data: any[]) => {
  if (DEBUG >= level) console.debug(...data);
};

const input = await Deno.readTextFile(FILENAME);

const parseInput = () => {
  const grid = input.split('\n').filter((line) => line.trim());
  const rows = grid.length;
  const cols = grid[0].length;
  return { grid, rows, cols };
};

const part1 = () => {
  const { grid, rows, cols } = parseInput();
  const partRegex = /[0-9]+/g;
  const gearRegex = /[^0-9.]/;
  const partNumbers: number[] = [];
  for (const [rowId, row] of grid.entries()) {
    debug(1, { row, rowId });
    for (const match of row.matchAll(partRegex)) {
      const startIx = match.index;
      const endIx = match.index + match[0].length;
      const partNumber = Number(match[0]);
      const checkStartIx = startIx > 0 ? startIx - 1 : startIx;
      const checkEndIx = endIx < cols ? endIx + 1 : endIx;
      const surroundings: Map<string, string> = new Map();
      if (rowId > 0) surroundings.set('above', grid[rowId - 1].substring(checkStartIx, checkEndIx));
      if (rowId < rows - 1) surroundings.set('below', grid[rowId + 1].substring(checkStartIx, checkEndIx));
      if (checkStartIx < startIx) surroundings.set('left', row.substring(checkStartIx, startIx));
      if (checkEndIx > endIx) surroundings.set('right', row.substring(endIx, checkEndIx));
      const gearMatch = gearRegex.test(Array.from(surroundings.values()).join(''));
      debug(1, { startIx, endIx, partNumber, checkStartIx, checkEndIx, surroundings, gearMatch });
      if (gearMatch) partNumbers.push(partNumber);
    }
  }
  debug(1, partNumbers);

  const sum = partNumbers.reduce((acc, item) => acc + item, 0);
  console.log('part 1', sum);
};

const part2 = () => {
  interface Gear {
    r: number;
    c: number;
  }

  const { grid, rows, cols } = parseInput();
  const partRegex = /[0-9]+/g;
  const gearRegex = /\*/g;

  const findGears = (rowId: number, colIdStart: number, colIdEnd: number) => {
    const result: Gear[] = [];
    if (0 <= rowId && rowId < rows && colIdStart < colIdEnd) {
      for (const gearMatch of grid[rowId].substring(colIdStart, colIdEnd).matchAll(gearRegex)) {
        result.push({ r: rowId - 1, c: colIdStart + gearMatch.index });
      }
    }
    return result;
  };

  const gears: Map<string, number[]> = new Map();
  for (const [rowId, row] of grid.entries()) {
    debug(1, { row, rowId });
    for (const match of row.matchAll(partRegex)) {
      const startIx = match.index;
      const endIx = match.index + match[0].length;
      const partNumber = Number(match[0]);
      const checkStartIx = startIx > 0 ? startIx - 1 : startIx;
      const checkEndIx = endIx < cols ? endIx + 1 : endIx;
      const surroundingGears: Gear[] = [];
      surroundingGears.push(
        ...findGears(rowId - 1, checkStartIx, checkEndIx),
        ...findGears(rowId + 1, checkStartIx, checkEndIx),
        ...findGears(rowId, checkStartIx, startIx),
        ...findGears(rowId, endIx, checkEndIx)
      );
      debug(1, { startIx, endIx, partNumber, checkStartIx, checkEndIx, surroundingGears });
      if (surroundingGears.length == 1) {
        // ugh
        const gearUnique = JSON.stringify(surroundingGears[0]);
        gears.set(gearUnique, [...(gears.get(gearUnique) ?? []), partNumber]);
      }
    }
  }
  debug(1, gears);

  const sum = Array.from(gears.values())
    .filter((v) => v.length == 2)
    .reduce((acc, [left, right]) => acc + left * right, 0);
  console.log('part 2:', sum);
};

if (PART1) part1();
if (PART2) part2();
